source/custom/visa_service.custom.cpp

```diff
@@ -0,0 +1,440 @@
+#include <visa/visa_service.h>
+
+namespace visa_grpc {
+using nidevice_grpc::converters::convert_from_grpc;
+using nidevice_grpc::converters::convert_to_grpc;
+
+static ViSession s_defaultRM = VI_NULL;
+static std::shared_mutex s_resource_manager_lock;
+
+// Returns true if it's safe to use outputs of a method with the given status.
+inline bool status_ok(int32 status)
+{
+  return status >= VI_SUCCESS;
+}
+
+class DriverErrorException : public std::runtime_error
+{
+  private:
+    int status_ = 0;
+
+  public:
+    DriverErrorException(int status) : std::runtime_error(""), status_(status)
+    {
+    }
+    int status() const
+    {
+      return status_;
+    }
+};
+
+static ViSession GetResourceManagerSession(visa_grpc::VisaService::LibrarySharedPtr library)
+{
+    std::unique_lock<std::shared_mutex> lock(s_resource_manager_lock);
+    if (!s_defaultRM)
+    {
+        ViStatus status = library->OpenDefaultRM(&s_defaultRM);
+        if (!status_ok(status))
+        {
+            throw DriverErrorException(status);
+        }
+    }
+    return s_defaultRM;
+}
+
+::grpc::Status VisaService::ConvertApiErrorStatusForViSession(::grpc::ServerContextBase* context, int32_t status, ViSession vi)
+{
+  std::string description(nidevice_grpc::kMaxGrpcErrorDescriptionSize, '\0');
+  library_->StatusDesc(vi, status, &description[0]);
+  return nidevice_grpc::ApiErrorAndDescriptionToStatus(context, status, description);
+}
+
+::grpc::Status VisaService::ConvertApiErrorStatusForViObject(::grpc::ServerContextBase* context, int32_t status, ViObject vi)
+{
+  std::string description(nidevice_grpc::kMaxGrpcErrorDescriptionSize, '\0');
+  library_->StatusDesc(vi, status, &description[0]);
+  return nidevice_grpc::ApiErrorAndDescriptionToStatus(context, status, description);
+}
+
+//---------------------------------------------------------------------
+  //---------------------------------------------------------------------
+  ::grpc::Status VisaService::Close(::grpc::ServerContext* context, const CloseRequest* request, CloseResponse* response)
+  {
+    if (context->IsCancelled()) {
+      return ::grpc::Status::CANCELLED;
+    }
+    try {
+      ViStatus status;
+      if (request->object_handle().has_vi())
+      {
+        auto vi_grpc_session = request->object_handle().vi();
+        ViSession vi = session_repository_->access_session(vi_grpc_session.name());
```

Comment:
Declare the local variable as `ViObject vi` above the if clause. The lines to actually call `Close()` and check the status can move below the if/else. That way the only part inside the conditional is how it gets assigned and what happens (one block removes the session, the other does not).

Reply:
Dan has an elegant implementation for Close in his new commit in the draft PR, and will post a PR with the set/get attributes implementation after this.